Sources/TuistMigration/Utilities/TargetsExtractor.swift

```diff
@@ -0,0 +1,83 @@
+import Foundation
+import PathKit
+import TSCBasic
+import TuistSupport
+import XcodeProj
+
+/// An interface to extract all targets from an xcode project, sorted by number of dependencies
+public protocol TargetsExtracting {
+    /// - Parameters:
+    ///   - xcodeprojPath: Path to the Xcode project.
+    func targetsSortedByDependencies(xcodeprojPath: AbsolutePath) throws -> [(targetName: String, dependenciesCount: Int)]
+}
+
+public final class TargetsExtractor: TargetsExtracting {
+    // MARK: - Init
+
+    public init() {}
+
+    // MARK: - EmptyBuildSettingsChecking
+
+    public func targetsSortedByDependencies(xcodeprojPath: AbsolutePath) throws -> [(targetName: String, dependenciesCount: Int)] {
+        guard FileHandler.shared.exists(xcodeprojPath) else { throw TargetsExtractorError.missingXcodeProj(xcodeprojPath) }
+        let pbxproj = try XcodeProj(path: Path(xcodeprojPath.pathString)).pbxproj
+        let targets = pbxproj.nativeTargets + pbxproj.aggregateTargets + pbxproj.legacyTargets
+        if targets.isEmpty {
+            throw TargetsExtractorError.noTargets
+        }
+        do {
+            return try sortTargetsByDependenciesCount(targets)
+        } catch {
+            throw TargetsExtractorError.failedToExtractTargets(error.localizedDescription)
+        }
+    }
+
+    private func sortTargetsByDependenciesCount(_ targets: [PBXTarget]) throws -> [(targetName: String, dependenciesCount: Int)] {
+        let sortedTargets = try targets.sorted { lTarget, rTarget -> Bool in
+            let lCount = try countDependencies(of: lTarget)
+            let rCount = try countDependencies(of: rTarget)
+            if lCount == rCount {
+                return lTarget.name < rTarget.name
+            }
+            return lCount < rCount
+        }
+        return try sortedTargets.map { (targetName: $0.name, dependenciesCount: try countDependencies(of: $0)) }
+    }
+
+    private func countDependencies(of target: PBXTarget) throws -> Int {
+        var count = target.dependencies.count
+        if let frameworkFiles = try target.frameworksBuildPhase()?.files {
+            count += frameworkFiles.count
+        }
+        return count
+    }
+}
+
+enum TargetsExtractorError: FatalError, Equatable {
```

Comment:
There's an undocumented convention of placing the errors at the top of the file.

Reply:
👍